### PR TITLE
Fix linting rules for allowing a watchOS as a dependency of a test target

### LINKED ISSUE #4907: Linting error when a Xcode 14 watchOS .application is added as a dependency in unit test target
### What happened?

When declaring a `dependency` to a `.application` target with platform `.watchOS` inside a `.unitTest` product the following error message is shown: 

```
Target FrameworkTests_watchos has platform 'watchOS' and product 'unit tests' and depends on target watchOSApp of type application and platform 'watchOS' which is an invalid or not supported yet combination.
```
The way how it is configured is the following: 

```swift
let frameworkWatchOSUnitTestTarget = Target(
    name: "FrameworkTests_watchos",
    platform: .watchOS,
    product: .unitTests,
    bundleId: "com.test.framework.tests",
    sources: ["Framework/Tests/**/*.swift"],
    dependencies: [
        .target(name: "Framework_watchos"),
        .target(name: "watchOSApp") // this does not work !
    ]
)
```

For a `.unitTest` product with platform `.iOS` this seems to be not a problem. 

```swift

let frameworkiOSUnitTestTarget = Target(
    name: "FrameworkTests_ios",
    platform: .iOS,
    product: .unitTests,
    bundleId: "com.test.framework.tests",
    sources: ["Framework/Tests/**/*.swift"],
    dependencies: [
        .target(name: "Framework_ios"),
        .target(name: "iOSApp") // this works
    ]
)
```

I assume its because Xcode 14 added support for a standalone watch app without accompanying extension target, where support was added here https://github.com/tuist/tuist/pull/4658 but linting rules where not adapted here: https://github.com/tuist/tuist/blob/948c435c75891458ef28f4055659a0d11ab623b3/Sources/TuistGenerator/Linter/GraphLinter.swift#L534

But I am not 100% sure, though. 

### How do we reproduce it?

I have created an example project which produces the error mentioned above: 

* Checkout https://github.com/alexanderwe/tuist-watchos-unit-test-host-app-issue
* Run `tuist generate`
* See the linting error

### Error log

```sh
Target FrameworkTests_watchos has platform 'watchOS' and product 'unit tests' and depends on target watchOSApp of type application and platform 'watchOS' which is an invalid or not supported yet combination.
```

### macOS version

13.0

### Tuist version

3.14.0

### Xcode version

14.1.0

### PR DESCRIPTION
**Pull request by alexanderwe**
Resolves https://github.com/tuist/tuist/issues/4907

### Short description 📝

This PR should fix #4907 by enabling the combination of `product: .app, platform: .watchOS` to be valid when included in an `.unitTest` and `.uiTest` target. This should fix issues when the new way of using a watchOS app target, introduced in this PR https://github.com/tuist/tuist/pull/4658 is used. 

I am not sure how we should handle older Xcode versions here, but I guess it should not be possible to generate a project with only a `product: .app, platform: .watchOS` target in, with Xcode versions lower than 14.0. But I would love to hear your opinion about that. 

### How to test the changes locally 🧐

1. Checkout  https://github.com/alexanderwe/tuist-watchos-unit-test-host-app-issue
2. Run `tuist generate`
3. Verify generation works 
4. Verify watchOS host app is added correctly to `FrameworkTests_watchos` target 
5. Verify tests are running 

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] The title of the PR will be used as changelog entry, please make sure it is clear and suitable
- [ ] In case the PR introduces changes that affect users, the documentation has been updated
- [ ] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`

### CODE REVIEW

**Comment by alexanderwe** [comment]
@kwridan Very weird behaviour. When I clean the build directory, tests are working fine. As soon as I re-generate the project I get this cyclic dependency issue. Do you maybe have an idea why this could happen ? Never saw it before. 

1. Generate project for change fixture or my example project 
2. Run tests for watchOS 
3. Works 
4. Close Xcode 
5. Re-generate the project 
6. Try to run tests again
7. Xcode build failure

File: projects/tuist/fixtures/ios_app_with_watch_application/FrameworkA/Tests/FrameworkATests.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import XCTest
+@testable import FrameworkA
+
+class FrameworkATest: XCTestCase {
+
+	func testProperty() async throws {
+        XCTAssertEqual(FrameworkA.text, "Hello, FrameworkA!")
+	}
+
```
Comment (danyf90) [inline projects/tuist/fixtures/ios_app_with_watch_application/FrameworkA/Tests/FrameworkATests.swift:9]:
```suggestion
    func testProperty() async throws {
        XCTAssertEqual(FrameworkA.text, "Hello, FrameworkA!")
    }
```

**Comment by kwridan** [comment]
Thanks for the updates @alexanderwe and good thing we went ahead with adding this fixture as it did reveal a bug / missing scenario we need to handle (that was previously never possible due to the lint rules!)

The generated test target now also embed the watch application which does indeed cause a cycle. This is not expected - the embedding of watch applications should only take place within host iOS applications 

watchapp embedding

We'll need to update https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift#L496 to only add this extra build phase to targets that support embedding a watching application (to my knowledge this is currently limited to iOS application targets)

We can add another property on `Target` `canHostWatchApplications` which can help us filter / limit where this build phase gets added.


Minor note, I would have imagined the fixture update including a `WatchAppTests` target - was this not the intention?

```swift
        Target(
            name: "WatchAppTests",
            platform: .watchOS,
            product: .unitTests,
            bundleId: "io.tuist.App.watchkitapptests",
            infoPlist: .default,
            sources: "WatchApp/Tests/**",
            dependencies: [
                .target(name: "WatchApp"),
            ]
        ),
```

Thanks again @alexanderwe

**Comment by alexanderwe** [comment]
> Thanks for the updates @alexanderwe and good thing we went ahead with adding this fixture as it did reveal a bug / missing scenario we need to handle (that was previously never possible due to the lint rules!)
> 
> The generated test target now also embed the watch application which does indeed cause a cycle. This is not expected - the embedding of watch applications should only take place within host iOS applications
> 
> watchapp embedding
> 
> We'll need to update https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift#L496 to only add this extra build phase to targets that support embedding a watching application (to my knowledge this is currently limited to iOS application targets)
> 
> We can add another property on `Target` `canHostWatchApplications` which can help us filter / limit where this build phase gets added.
> 
> Minor note, I would have imagined the fixture update including a `WatchAppTests` target - was this not the intention?
> 
> ```swift
>         Target(
>             name: "WatchAppTests",
>             platform: .watchOS,
>             product: .unitTests,
>             bundleId: "io.tuist.App.watchkitapptests",
>             infoPlist: .default,
>             sources: "WatchApp/Tests/**",
>             dependencies: [
>                 .target(name: "WatchApp"),
>             ]
>         ),
> ```
> 
> Thanks again @alexanderwe

Oh I must have overseen this fact ! Thanks for pointing it out. I will take a look at the generator and try to implement this check and also see what Xcode is doing when I add the host app there. 

My mistake for the fixture. I thought it should be similar to the example repo by me. There it was the framework tests needing the host app. I can also add the tests you suggested

File: projects/tuist/fixtures/ios_app_with_watch_application/WatchApp/Tests/WatchTests.swift [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import XCTest
+
```
Comment (kwridan) [inline projects/tuist/fixtures/ios_app_with_watch_application/WatchApp/Tests/WatchTests.swift:2]: nit: we can do something like

```suggestion
@testable import WatchApp
```

And attempt to access a type defined within in the test for completeness

**Review by kwridan — APPROVED** [review]
Awesome, thank you for all the updates, it's looking great 👍

File: projects/tuist/fixtures/ios_app_with_watch_application/WatchApp/Tests/WatchTests.swift (hunk 2 above)
Comment (alexanderwe) [reply projects/tuist/fixtures/ios_app_with_watch_application/WatchApp/Tests/WatchTests.swift:3]: Good point, added

**Comment by alexanderwe** [comment]
just saw `test_observable_when_it_errors` failed in the tests. Do you think it's related to my changes ? I couldn't see why the changes influence it but I might be wrong. Locally the test was working for me

**Comment by alexanderwe** [comment]
Thanks @danyf90 for merging `main` in. Tests are green now. 
Anything else I need to take care of @kwridan ?

**Review by danyf90 — APPROVED** [review]

**Comment by kwridan** [comment]
> Anything else I need to take care of @kwridan ?

Awesome, thanks for the follow ups on this one!

**Comment by alexanderwe** [comment]
@kwridan, @danyf90  Thanks a lot for reviewing and taking care :)

**Pull request merged**